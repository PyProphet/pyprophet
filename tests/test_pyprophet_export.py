from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout


def _run_osw(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # peptide-level
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=global"

    # protein-level
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=global"

    # export
    cmdline += " export --in=test_data.osw --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    if not peptide:
        cmdline += " --no-peptide"

    if not protein:
        cmdline += " --no-protein"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_osw_unscored(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # export
    cmdline = "pyprophet export --in=test_data.osw --out=test_data.tsv --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_ipf(regtest, temp_folder, transition_quantification=False, ipf="disable"):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # MS2-level
    cmdline += " score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # IPF
    cmdline += " ipf --in=test_data.osw"

    # export
    cmdline += " export --in=test_data.osw --no-peptide --no-protein --ipf_max_peptidoform_pep=1  --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    cmdline += " --ipf=" + ipf

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_unscored(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # export
    cmdline = "pyprophet export-compound --in=test_data_compound.osw --out=test_data_compound_unscored.tsv --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_unscored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_compound_ms1(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms1.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms1.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_ms2(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS2-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms2 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms2.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms2.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)


def test_osw_0(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, False)

def test_osw_1(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, True, False, False)

def test_osw_2(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, True, False)

def test_osw_3(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, True)

def test_ipf_0(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "disable")

def test_ipf_1(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, True, "disable")

def test_ipf_2(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "peptidoform")

def test_ipf_3(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "augmented")

def test_compound_0(tmpdir, regtest):
    _run_compound_ms1(regtest, tmpdir.strpath)
    
def test_compound_1(tmpdir, regtest):
    _run_compound_ms2(regtest, tmpdir.strpath)

def test_compound_unscored(tmpdir, regtest):
	_run_compound_unscored(regtest, tmpdir.strpath)
	
def test_osw_unscored(tmpdir, regtest):
	_run_osw_unscored(regtest, tmpdir.strpath)
