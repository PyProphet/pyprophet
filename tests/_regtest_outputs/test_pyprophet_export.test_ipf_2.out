    Charge                  FullPeptideName  Intensity              ProteinName       RT             Sequence  aggr_prec_Peak_Apex  aggr_prec_Peak_Area  assay_iRT   assay_rt  d_score  decoy  delta_iRT  delta_rt                         filename      iRT                   id     ipf_pep  leftWidth     m_score  ms1_pep  ms2_m_score  ms2_pep        mz  peak_group_rank  precursor_pep  rightWidth               run_id                  transition_group_id
0        2           AAEDFTLLVK(UniMod:259)   189687.0      AQUA4SWATH_HMLangeE  3665.82           AAEDFTLLVK          107855.1703             702922.0       58.9  3708.4232   5.5390      0    -1.3421  -42.6032  napedro_L120420_010_SW.mzXML.gz  57.5579  6227384657721288232  0.0000e+00  3646.8899  0.0000e+00   0.0003       0.0033   0.0031  557.8153                1     9.4030e-09   3698.1001 -8670811102654834151            AAEDFTLLVK(UniMod:259)_87
1        2       AAGASAQVLGQEGK(UniMod:259)   265021.0          AQUA4SWATH_Spyo  1521.11       AAGASAQVLGQEGK          231000.0046            1051340.0       -5.3  1503.3343   5.8582      0     0.6697   17.7757  napedro_L120420_010_SW.mzXML.gz  -4.6303 -4109405113780929799  0.0000e+00  1503.2700  0.0000e+00   0.0003       0.0033   0.0031  647.8457                1     7.0822e-10   1537.4000 -8670811102654834151       AAGASAQVLGQEGK(UniMod:259)_257
2        2    AASEIATAELAPTHPIR(UniMod:267)    63749.0  AQUA4SWATH_PombeSchmidt  2754.99    AASEIATAELAPTHPIR           56607.9433             320548.0       31.5  2767.3108   5.6965      0    -0.3526  -12.3208  napedro_L120420_010_SW.mzXML.gz  31.1474 -4579089096808240748  0.0000e+00  2736.7000  0.0000e+00   0.0003       0.0033   0.0031  879.4746                1     7.0822e-10   2774.2600 -8670811102654834151    AASEIATAELAPTHPIR(UniMod:267)_235
3        2         AAVPVLVHPAPR(UniMod:267)    47860.0  AQUA4SWATH_Tuberculosis  2370.16         AAVPVLVHPAPR           30508.2548             135628.0       19.6  2358.5846   5.1771      0     0.3887   11.5754  napedro_L120420_010_SW.mzXML.gz  19.9887  1080224756928033227  0.0000e+00  2349.8101  0.0000e+00   0.0003       0.0033   0.0031  618.8762                1     1.5983e-09   2390.7800 -8670811102654834151         AAVPVLVHPAPR(UniMod:267)_279
4        2           ADDYTYEHLR(UniMod:267)    97468.0      AQUA4SWATH_HMLangeG  1743.79           ADDYTYEHLR           89956.6100             330373.0        4.8  1850.2388   5.3371      0    -2.9734 -106.4488  napedro_L120420_010_SW.mzXML.gz   1.8266 -6595106422864423511  0.0000e+00  1728.5800  0.0000e+00   0.0003       0.0033   0.0031  646.7927                1     7.0822e-10   1759.3101 -8670811102654834151           ADDYTYEHLR(UniMod:267)_132
..     ...                              ...        ...                      ...      ...                  ...                  ...                  ...        ...        ...      ...    ...        ...       ...                              ...      ...                  ...         ...        ...         ...      ...          ...      ...       ...              ...            ...         ...                  ...                                  ...
95       2            GFDATYHVR(UniMod:267)    92840.0      AQUA4SWATH_HMLangeE  1750.75            GFDATYHVR          135806.9267            1026610.0        3.9  1819.3275   4.4511      0    -1.8716  -68.5775  napedro_L120420_010_SW.mzXML.gz   2.0284  6575595963013729151  3.9957e-13  1731.5900  1.8144e-15   0.5778       0.0062   0.0225  538.2634                1     8.4453e-03   1769.1400 -8670811102654834151             GFDATYHVR(UniMod:267)_95
96       2  GFTPDGPAGTEPNIAPGYK(UniMod:259)   304459.0         AQUA4SWATH_Lepto  3075.75  GFTPDGPAGTEPNIAPGYK          194847.4074             968985.0       39.4  3038.6533   5.9861      0     1.0482   37.0967  napedro_L120420_010_SW.mzXML.gz  40.4482   725624140668631730  0.0000e+00  3057.8201  0.0000e+00   0.0003       0.0033   0.0031  948.9645                1     7.0822e-10   3095.3701 -8670811102654834151  GFTPDGPAGTEPNIAPGYK(UniMod:259)_169
97       2           GGNFGFGDSR(UniMod:267)   382433.0  AQUA4SWATH_HumanEbhardt  1984.57           GGNFGFGDSR          158928.4812             843945.0       11.9  2094.1085   5.0378      0    -3.0918 -109.5385  napedro_L120420_010_SW.mzXML.gz   8.8082 -7797929030476974532  0.0000e+00  1963.6300  0.0000e+00   0.0003       0.0033   0.0031  512.2295                1     5.9380e-08   2011.4301 -8670811102654834151           GGNFGFGDSR(UniMod:267)_149
98       2          GHFYYDISDVR(UniMod:267)    55289.0      AQUA4SWATH_HMLangeG  2889.39          GHFYYDISDVR           46057.1329             272430.0       39.4  3038.6590   4.9333      0    -4.3557 -149.2690  napedro_L120420_010_SW.mzXML.gz  35.0443  9218597765662578237  0.0000e+00  2869.0300  0.0000e+00   0.0003       0.0033   0.0031  691.3242                1     2.4091e-08   2913.4099 -8670811102654834151          GHFYYDISDVR(UniMod:267)_134
99       2         GIAASDGVAVAK(UniMod:259)   319128.0          AQUA4SWATH_Spyo  1715.56         GIAASDGVAVAK          228311.6217            1060010.0       -1.4  1637.2919   5.8413      0     2.4079   78.2681  napedro_L120420_010_SW.mzXML.gz   1.0079  7617756229621228830  0.0000e+00  1697.4500  0.0000e+00   0.0003       0.0033   0.0031  533.8028                1     7.2449e-09   1735.0000 -8670811102654834151         GIAASDGVAVAK(UniMod:259)_263

[100 rows x 29 columns]
